acvm/src/compiler.rs

```diff
@@ -68,23 +68,28 @@ pub fn compile(
     let mut transformed_gates = Vec::new();
 
     let mut next_witness_index = acir.current_witness_index + 1;
+    let mut intermediate_variables: IndexMap<Expression, (FieldElement, Witness)> = IndexMap::new();
     for opcode in acir.opcodes {
         match opcode {
             Opcode::Arithmetic(arith_expr) => {
-                let mut intermediate_variables: IndexMap<Witness, Expression> = IndexMap::new();
+                let len = intermediate_variables.len();
 
                 let arith_expr = transformer.transform(
                     arith_expr,
                     &mut intermediate_variables,
-                    next_witness_index,
+                    &mut next_witness_index,
                 );
 
                 // Update next_witness counter
-                next_witness_index += intermediate_variables.len() as u32;
+                next_witness_index += (intermediate_variables.len() - len) as u32;
                 let mut new_gates = Vec::new();
-                for (_, mut g) in intermediate_variables {
-                    g.sort();
-                    new_gates.push(g);
+                for (g, (l, w)) in intermediate_variables.iter().skip(len) {
+                    // de-normalise
+                    let mut new_g = Expression::default().add_mul(*l, g);
+                    // constrain the intermediate gate to the intermediate variable
+                    new_g.linear_combinations.push((-FieldElement::one(), *w));
```

Comment:
This is going to call the sub() function, then create an expression from the witness, so a bunch a vectors, then it is going to call another sub(), which is finally calling add_mul() who is going to iterate over the vectors and perform some field multiplication(s)...
So in fact adding directly the linear term is better, and because w is a new variable, we know it is safe to do so we can avoid the 'merge process' done in add_mul().